### ISSUE TITLE
Error when calculating the inverse matrix

### ISSUE THREAD

**Issue by JihyoKim00** [jbonato1/SCAR#1]
Thanks for the interesting work!

While trying to reproduce SCAR results using this repository, the error `torch._C._LinAlgError: linalg.svd: The algorithm failed to converge because the input matrix is ill-conditioned or has too many repeated singular values (error code: 511)` was raised when calculating the pseudo inverse matrix.

https://github.com/jbonato1/SCAR/blob/a725b3168e76e35ae943108b27dfeada874a13d8/src/Unlearning_methods.py#L377

I used the same script as the sample script you provided for CIFAR 10 in CR scenario.

**Comment by jbonato1** [comment]
Can you give us a description of what you did?  i.e. config parameters, which resnet weights ecc.
Is the cov matrix shrinked ok? Did you look at it?

**Comment by JihyoKim00** [comment]
I’m trying to unlearn one specific CIFAR10 class on ResNet18 model with CIFAR10 dataset in CR scenario.
I used weights from the original model which was trained on the entire CIFAR10 using `training_original.py`, and checked that the original model's accuracy was fine.
The hyperparameters I used were batch size 1024, learning rate 0.0005, temperature 1 lambda1  1, lambda2 5, delta 0.5, gamma1 3, and gamma2 3 following the example script.
I've checked the shrinked cov matrix but haven't found any unexpected values such as NaN.

**Comment by jbonato1** [comment]
Ok I'll try to reproduce this error from the released github code

**Comment by jbonato1** [comment]
I 've found some bugs on the code. I fixed them and pushed the code. Now cifar10 CR scenario works on the docker image released. Let me know

**Comment by JihyoKim00** [comment]
Thanks for checking fast!

I actually had the same error with the fixed code, but I think I've found the cause. 
https://github.com/jbonato1/SCAR/blob/280ceaebd69e81f1619cf627950bb7b95e0104aa/src/Unlearning_methods.py#L363-L367

In these lines, filtering forget class only works if the forget class is provided in the nested list form, but I was trying to unlearn only one class, so I used the list not in the form of the nested list. After changing the form of `class_to_remove` in opt.py to the nested list even if the forget class is only one, it works well. Sorry about the confusion.

But my reproduced SCAR results on the retain test accuracy was quite lower than the reported, which is closed to 60%. I used `surrogate_ImageNet.txt` for the surrogated dataset. Could you kindly let me know if you've had any reproducibility issues with the current code? If everything is working on your end, I'll double-check mine.

**Comment by jbonato1** [comment]
Ok its strange...I'm getting the correct results :)  Could you please send the screenshots of the outputs on terminal from the beginning till the forget_test_accuracy  and retain test acc results?  So i can understand whats happening?
What is the test accuracy of the original model that you trained?

In main.py you can comment lines 113 and 114 and use line  115
```python
        elif opt.mode == "CR":
            #df_un_model = get_MIA_SVC(train_loader=None, test_loader=test_loader,model=unlearned_model,opt=opt,fgt_loader=train_fgt_loader,fgt_loader_t=test_fgt_loader)
            #print('F1 mean: ',df_un_model.F1.mean())
            df_un_model = pd.DataFrame([0],columns=["PLACEHOLDER"])
```

**Comment by JihyoKim00** [comment]
The original model is reset18 trained on CIFAR10 and its accuracy on CIFAR10 test set is 84.72%. It is slightly below 88% which is reported in the paper but I thought it would not be a problem.

This is a script I used for SCAR 
`python3 main.py --run_name cifar10_CR --dataset cifar10 --mode CR --save_model --save_df --run_unlearn  --num_workers 4 --method SCAR --model resnet18 --bsize 1024 --lr 0.0005 --epochs 30  --temperature 1 --lambda_1 1 --lambda_2 5 --delta .5 --gamma1 3 --gamma2 3`

and the below is a screenshot of my training logs. I cropped it from the unlearning process for class 9 because the logs for each class are similar.
![Screenshot 2024-10-18 at 14 06 24](https://github.com/user-attachments/assets/ecf8142b-6943-4d03-a891-9ead2794610f)

Since you have no issues and if there are no noticeable points in my logs, there may be some errors in mine. Please let me know if there are some strange points!

**Comment by jbonato1** [comment]
Thanks for the information! I’ll look into the issue and try to figure out what’s going on. I’ll be busy until the 27th, so please hold off until the end of October. Thanks again!

**Comment by JihyoKim00** [comment]
Thank you so much while you are busy! I hope your work is going well.

**Comment by jbonato1** [comment]
I'm really sorry for this delay but maybe I've found the problem did you change    in opts.py num_retain_samp from 5 to 1?
